add queryable config parsing

jwmelto opened the pull request:
This request extends **pyzpl** to provide a queryable, hierarchical configuration file type interface. All existing tests pass, as well as the new one added for this capability.

jwmelto committed:
add queryable config parsing

jwmelto commented on pyzpl/__init__.py:
```diff
@@ -57,16 +57,19 @@
+else:                                     #python 3
+    # str supports necessary decode
+    vitems = operator.methodcaller('items')
+    
```
This is just preference, but with two version-specific  changes under V2 and only 1 under V3, I thought this was more straightforward

mbarkhau replied:
Sure 👍

jwmelto commented on pyzpl/__init__.py:
```diff
@@ -82,46 +85,70 @@
         if spaces % 4 != 0:
             msgfmt = "Illegal indent on line {}, must be a multiple of 4\n\t{}"
-            raise ValueError(msgfmt.format(spaces, repr(cleaned_line)))
+            raise ValueError(msgfmt.format(lineno, repr(cleaned_line)))
```
I'm pretty sure this fixes a bug

mbarkhau replied:
Looks like it 👍

jwmelto commented on pyzpl/__init__.py:
```diff
@@ -82,46 +85,70 @@
+            propname_stack += (nv_pair_match.group(1).strip(),)
             prev_indent_lvl = indent_lvl
-            yield propnames, value
+            yield propname_stack, nv_pair_match.group(2)
```
This definitely fixes a bug (where the "bar" items didn't have their "thing" parent specified). Also, it is stylistically consistent with the handling of `propname_stack` below.

mbarkhau replied:
Seems to be the case yes. Conceptually this assumed that a name value pair would be a leaf node and so the propname_stack could be left untouched.

jwmelto commented on pyzpl/__init__.py:
```diff
@@ -82,46 +85,70 @@
                 propname_stack += (propname_match.group(1),)
                 prev_indent_lvl = indent_lvl
-
+                if emit_empty: yield propname_stack, ""
```
this adds the (optional) new behavior I needed, where every node is emitted.

jwmelto committed:
clean up cruft

mbarkhau commented on pyzpl/__init__.py:
```diff
@@ -82,46 +85,70 @@
         indent_lvl = spaces // 4
         if indent_lvl <= prev_indent_lvl:
             propname_stack = propname_stack[:indent_lvl]
+            # this represents an outdent?
```
Indeed it does.

mbarkhau commented:
The examples in the test are quite confusing. Perhaps it would be clearer if the property names were not the same as the property values, ie. `bar = bar`.

mbarkhau commented:
Would it make sense to always return only the value from the `ZPLnode.get` method?

mbarkhau commented on pyzpl/__init__.py:
```diff
@@ -180,6 +207,163 @@ def dumps(tree, *args, **kwargs):
+            self._level = 0
+        self._children = []
+
+    def get(self, path, filter=(None,), name_sep=':'):
```
`filter` is a python builtin function and it is bad practice to shadow it. It would be better to choose a different parameter name.

jwmelto replied:
Good point. I will change that

jwmelto replied:
> Would it make sense to always return only the value from the `ZPLnode.get` method?

No, because the result of `get()` can legitimately be a sub-tree. When using this interface, you are aware of the node structure and care about both the value and its children.

jwmelto committed:
review comments

mbarkhau commented:
I can't say that I like this API, and I still don't think the test code is particularly clear, but if you say this is as far as you're willing to take it, then I'll be happy to accept the contribution and release it as 0.2.0.

jwmelto commented:
> I can't say that I like this API, and I still don't think the test code is particularly clear, but if you say this is as far as you're willing to take it, then I'll be happy to accept the contribution and release it as 0.2.0.

I'm willing to entertain changes, but I need a bit more to go on. I'm accept the criticism of the test code, but the framework was designed for a simple transformation; the results of this parsing API is an in-memory "database" that can be queried for specific nodes/subtrees/etc. I wasn't sure how to fit useful tests into that framework.

mbarkhau commented:
If you're willing to do more with this, then I'm happy to give more concrete feedback. I'm just not sure what context you're in and I want to be respectful of your time.

Since we both agree on the tests, let's start there. The previous tests didn't cover querying in any way, they just validated that the round trip through parsing and serialization works (ie. `loads(data) -> tree` and `dumps(tree) -> data`. I think it would be better to have the tests for the query API separated out and not mixed together with the others.

The second thing about the tests is just about readability. Instead of "foo" and "bar", it would be better to have actual keys and values from an actual or theoretical use case. In particular when everything is a string, it should be clear which string is a key, which is a query and which is a value. For me it's always an indication that there is a problem when I feel the urge to write a clarifying comment as here:

~~~python
bar1 = cfg.get( ("thing", "bar"), query="bar" )   # get the bar of the any 'thing' node where the value is "bar"
~~~

Next let's consider the API itself. I think it would be good if the API were similar to existing APIs that a developer might be familiar with. For the case of pattern matching, I think the most well known APIs are CSS selectors or regular expressions. Now I know it's not Christmas anymore, but I've thought about what I think would be a nicer API.

~~~
# config file
node = frontend
    type = client
    bind = tcp://eth0:4444
node = backend_proxy
    type = proxy
    bind = tcp://eth0:6660
node = backend1
    type = backend
    bind = tcp://eth0:6661
node = backend2
    type = backend
    bind = tcp://eth0:6662
~~~


~~~python
cfg = pyzpl.load_cfg(...)

proxy_addr = cfg.get("node=backend_proxy:bind")
assert proxy_addr == "tcp://eth0:6660"

# same value but different query
proxy_addr = cfg.get("node[type=proxy]:bind")
assert proxy_addr == "tcp://eth0:6660"

# a query that doesn't match, just returns None
proxy_port = cfg.get("node[type=proxy]:port")
assert proxy_port is None

# for multiple matches the first one is returned
backend_addr = cfg.get("node[type=backend]:bind")
assert backend_addr == "tcp://eth0:6661"

node_ids = list(cfg.iter_all("node"))
assert node_ids == [
    "frontend",
    "backend_proxy",
    "backend1",
    "backend2",
]

backend_addrs = list(cfg.iter_all("node[type=backend]:bind"))
assert backend_addrs == [
    "tcp://eth0:6661",
    "tcp://eth0:6662",
]
~~~

It seems to me that in most cases a developer will want to query for values and not want to deal with nodes and subtrees themselves, but if they do want that, my guess is that the first method they would want to invoke would be `ZPLNode.to_dict(self)`.

~~~python
frontend_proxy = cfg.get_subtree("node=frontend_proxy")
assert frontend_proxy is None

backend_proxy = cfg.get_subtree("node=backend_proxy")
assert backend_proxy is not None
assert backend_proxy.to_dict() == {
    ":root": "backend_proxy",
    "type": "proxy",
    "bind": "tcp://eth0:6660",
}
~~~

I do want to thank you for your contribution, no matter how far you're willing to take it. I don't want you to be discouraged just because I have a wish list. I'll accept this PR whenever you consider it to be good enough for your needs (assuming the tests still pass), especially if you're waiting on this for some other project.

jwmelto commented:
I have a few questions.

First, how should I separate out the tests? I agree that they are awkward, but wasn't totally clear on how to extend the existing tests. I've used `pyunit` but not `pytest`.

Second, I'm not sure I'm following your query syntax. In particular `"node[type=proxy]:port"` seems especially odd. For ZPL, I don't know that it makes sense to query for a value based on the value of a sibling node.

I know you have used `:` as the default path element separator in describing ZPL, because it's not a valid name element. However, I don't know that it's necessarily a common one (I could be wrong). Note that your existing API returns a list of elements to represent hierarchy. It seemed to me that using a list of query params was consistent. Is it really that shocking? Or just a bias of your experience?

Third, while Python `dict` is an extremely useful construct, it's not necessarily appropriate for ZPL. Given that there is nothing that requires uniqueness of "keys", and given that Python doesn't have the equivalent of a `multimap`, I'm not convinced that `to_dict` is a reasonable interface.

I know you have a particular view of how ZPL would be accessed, which is evident in the API you already implemented. You said that I should add a new API if I needed a different access pattern (which I did). So I'm a bit confused that you seem to be pushing it back towards your existing construct?

I will say that the API I implemented does satisfy my needs, and is more congruent with the existing C API.

mbarkhau commented:
> First, how should I separate out the tests? 

It's quite simple, you just add a function with the prefix `test_` in `tests.py` and it will automatically be picked up. As an example see the `test_full_cycle` test function.

> Second, I'm not sure I'm following your query syntax. In particular "node[type=proxy]:port" seems especially odd. For ZPL, I don't know that it makes sense to query for a value based on the value of a sibling node.

Perhaps this is what you meant, but just to be clear in the query `node[type=proxy]:bind` what I intended to be selected is only value of the `bind` child of `node` which has a child `type` with the value `proxy`.

Perhaps the syntax is unclear. I was trying to base it on css selector syntax, but admittedly I was thinking of direct children as attributes of a node and was thinking of the [attribute selector syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors). 

> I know you have used : as the default path element separator in describing ZPL, because it's not a valid name element. However, I don't know that it's necessarily a common one (I could be wrong). 

You're right it's not very common, perhaps the space character would be a better choice.

> Note that your existing API returns a list of elements to represent hierarchy. It seemed to me that using a list of query params was consistent. Is it really that shocking? Or just a bias of your experience?

I don't understand, the existing API returns only a dictionary (in the case of `pyzpl.load`) and a generator that yields 2-tuples (in the case of `pyzpl.load_stream`). I also think that the tight positional connection between that `path` and `query` parameters is confusing.

> Third, while Python dict is an extremely useful construct, it's not necessarily appropriate for ZPL. Given that there is nothing that requires uniqueness of "keys", and given that Python doesn't have the equivalent of a multimap, I'm not convinced that to_dict is a reasonable interface.

This is a fair point and I'm not suggesting that it be the only API. I do think that many people generate ZPL data for which the dictionary is the natural choice. Before your ticket, I would have thought that no other choice even made any sense. 

> I know you have a particular view of how ZPL would be accessed, which is evident in the API you already implemented. You said that I should add a new API if I needed a different access pattern (which I did). So I'm a bit confused that you seem to be pushing it back towards your existing construct?

Are you really quite sure that ZPL was intended to be used that way, just because the C API allows for it? In the documentation for the [C-API](http://czmq.zeromq.org/czmq4-0:zconfig) the example is strict hierarchical, has unique keys between siblings and doesn't have both children and a value assigned to the same name. 

The thing that makes me feel uneasy is that you are using what is in my view a loophole or ambiguity in the spec and I'm trying to think of a more flexible API that would satisfy not only your particular use case, but also the more general use case of traversing a tree, which would be made possible by a selector syntax like css, xpath, [JsonPath](https://goessner.net/articles/JsonPath/) or [jq](https://stedolan.github.io/jq/), which is a more flexible approach than what you've implemented so far.

jwmelto committed:
Add subscripting access to children
Separate out new unit tests

jwmelto commented:
ok, I've added a single test for the new API. If you think the sub-conditions should each be separate tests, we can talk, but since they all have the precondition of a parsed config it made more sense to me to group them.

I've added "subscript" style traversal. It's a bit simpler than your example, but should satisfy the basic desire. It includes accessing a child by name (first matching child) or by name=value. Failure to match raises KeyError, like a `dict`. For more tolerant queries, I left the clunky `get()` interface.

>I don't understand, the existing API returns only a dictionary (in the case of pyzpl.load) and a generator that yields 2-tuples (in the case of pyzpl.load_stream). I also think that the tight positional connection between that path and query parameters is confusing.

The generator actually yields n-tuples, depending on the depth of the ZPL. These n-tuples are the "path" arguments to `get()`. The query argument is a tuple <= n (it is automatically left-filled with `None`, since `None` means don't query the value of that element). Ultimately there is a 1:1 correspondence between `path` elements and `query` elements. I grant that it is somewhat complex, but powerful interfaces frequently appear so, at first.

> Are you really quite sure that ZPL was intended to be used that way, just because the C API allows for it? In the documentation for the C-API the example is strict hierarchical, has unique keys between siblings and doesn't have both children and a value assigned to the same name.

However ZPL may have been _intended_ to be used, I submit that if the specification allows it, then it is both legal and moral to use it that way. There is little to be gained by arbitrarily limiting functionality.  The fact that I have a use case that takes advantage of what you accuse of being a loophole proves the point. And regardless of the examples, the C API does support this directly (since I'm already using it).

Barring any egregious faults with this last version, I think I'm done. Hopefully you find this submission to add value.

jwmelto committed:
minor documentation and robustness cleanup

mbarkhau commented:
Hi @jwmelto,

I'd like to merge this, but before I publish it to pypi, I will move your code to a module called `pyzpl2`. This is mainly to avoid breaking any projects that use the existing code. Both the old module and the new one will be installed by using `pip install pyzpl`.

If you're ok with this, I'd like to merge the PR. Is there any further work you want to do first?

jwmelto commented:
No, I'm good. Let me know when it's pushed.
